### ISSUE TITLE
CompoundPath children are lost after a Unite 

### ISSUE THREAD

**Issue by ghenania** [paperjs/paper.js#1281]
@lehni I have an interesting [use case ](http://sketch.paperjs.org/#S/nVlbb9y2Ev4rwr7EASSWtyE5LnqANjGQ4mRTx0nPS7cPrr2FjajrwHZcwEH+e78RqY0kK66OYyDaFYfDuc833M+r3elf29Xh6t2H7e3ZxapenV2dy/fNTv7uTq+r4+qHarf9u3px9dfHq0+78+PT24uDz39etu3Jp3Z7WG1W27vt7ur8fLOqq49YfHl6e3q4Wa1J12S9Ypu8u2ucZ5WcpdYyKWdddHVjIinNxK68DDHV+Z335i7v8Nq2zWQ5Gor36+CcCj7UjqyyVhuLQ1Sw0diLJirSms1dYwIpSz6AR9IqWeMI54agfMCX8tZxcPX+5V2/R9gkwybc9Xz77yZ4BRF9bCf72VJ5NzhJXookZc/92luC5Ja4JrbKJOtC65RxUAxbVPJEvpVnolRjQQdHbSMULoYaC9F4e792MIC2HpYMTsE6noSLZseZS9RBuARmYeKsdZmJHFNOARMjEnrNdbBBsSOX9kcLtWXD7WBXSCBohsLhvPu1MU7FRCbWpIOK+JA1wBkdG6FqixR12T1iI+fcr621kMEFsIlGeeKQxmzEEkM2YqkhG1kAmxSU9WRTzUZpAzb7TU2vQ8+217FXtpdOdPIqQm9EHdwUUuw0EBHq/si2OGNs4IFwcLbRipNJVDMruN92m0SEuvfG3m2zGoHgfk1WQ7woYeYQcDBmf/LQup1sI+sWDTMTyIVNSMgarBTHwHasUYm7EjF7s0wiRhuRAHuSgac5e6Qz3TDshvHRNgMNRXUYtzOpT4hn3yVeMmM238qBnk3OAUKuMhIwIe4SPlmy4yQoThrGx4hN1snqqJynUFOQYhK9GyXBPnoHhhjZNyeB03C10cjrYKKK2pnMxmtT9+krbEBuZpMgu8lFFD6dIgysVSBy4xz494jBMfdrTsowxPWMYojiQ2ON+kR6XCPrGL7xDjQmqQhFRom0d1K/KaY4TiQEA3zttHKGErgkh6wG86ewQfUU3wRbU0KZgC4TNn3kDVQYs8m+dkmRN6Ah2Dca1JuRaboQGSaTsH2YTERQKgZjaoJprHwcF/FeGjwfk4akoVmDnNSwdSI9CeCJbfZJOcx1xIx0SHaoVQnBwzrE2QAeFrixu8FNOoqFMGwsOkpQJj6RjU9IRYrWI7chFhq1HUVw31P2Eby38CgP0GoVvI3WRGjxzNb4cTqVRNin03zpdD6qwFp7KJUUNpc+Oa3jw040LhG5G0AVRpdFCY4eqRWewoS0R+whiGvEnyJOCJq5tBzGyEy3dSxYIqHbBnQXdH47bnB7aaaZMJYGBQ+VPJFDt3ViIz9RalKDHyq173CaghgYrcG7KRDZt9uSCZ3fmon/JaEMQJ6JFlXC4hOl2dh7ULFKt8qW8R6AKsIQKFhwu4GRltW9ERdoKtkEjVA5ldaAoeMS0ZfgkoPYPFM84RVFiUUh9CaXBBv2ZpgrNF+hlUT4ng05AgwBoETgAfCxJjNKg2k2fWUzqDzQSVvBRFxbdCk0qRSyD8iF7IMiXYIwnVDI/X69KQSIPABaBwgscBEtE72muDKFzMZzdhuedXnf9utNWYBOyaMueEGzLJ2SvckiB0Rw1wEKNMGzLu/bfr0pC5lNiKxd1+VYW9s239o15dofl3XSIQF0BhRy1skW82UnRJ99jWc2VDZ/NEUQL6UqJkBxDRTikY82EP/fLBwQHvmIgSZKSnchNb9nyrGcJNUF0NtINJFHiUK8FcgEX9bFt7M+lvWmLOSWJOAQbJIUCApFm7ltU7b9eQLCo6SiNBMt4xsGqKewQdwptEekJxIGCA8KPkkag9KkZU5DOjkZVebjlmLJGY4zcSvCkMR/BxSDtiE8hQ2APDpbRLFg6Ug+FSYxh21Mac7XstyUdZkIUCThY4jivEooUfwNLqlkYrQzbAR9J4AqlDowRD5zXzxmtk3Z9ueBDabbYBj1UQYUY2xyT5EGWYwWIEUU5sVQoG03e3sdL8ozD/n4MCD2KNIpYJ5eQmwBkDhgNlxC7FDCLfKI91cAI2K8GIoBPwCl6LSEGF/lPiT6RZxBbEG8SAyHZoOCGuMiYgQgRlJyi8SA6eAYu0wMgGtiYIpFxKhXjjBVLjIdOrsH8LGLZCbpmSku4wxiwFZaZDqLQkJw92Oc0f5lDnQRYYeyo20cB3Tsb626sBtQB6mdJi4iFjQI6LaQ2ueeycsEcRQxi5IPy3h7lDMHoL6M2ghWoxBGBtxTTw0I6ghqs4zas9JueHf3GDWqHmxi2S6i9jLOoqukZbytkXGTF8oN3kAghpZJAuRCUVu3TBI0LxbUtYxaY1QHdHyM2mNCQwMxMqE55QC2bRlx0U+6NugzasGzLu/bfr0pC2geKBWYiRjNGWwSBqR+Ug5lWikDZCwzsCDcfr1gAZGG4EUtjdgijQkjbEHT3KE/B9Talmfp1eV6hjOmlAW5UUqyuRtDEM0UI4/QNKYC35bnfngdwXosSIuH27EZIIoBLyGxKwjX5OtYzf1NZkYOgTMbWW8KgYB/4FprrbRnkhQqV4xGZ1TImgvYscXARUddcCEIgC2TQWdCJGJeRHMIOjyFi5GebLwXxIFRDYiZ2n61LuTtnNtkvT8PwpCWkmLEMhg6XTYv6UTZLv1FYDcTEYavtl9uyrqgFg0A1cF6lBzCqPY1vUjuMFATh1GNkccRGug8OXsm85V8s/ry/PvN7ljdnJ222wO0sbrq/pOfQY6vLne3B7rWz4VGfiB5++QfSEgu3Lvb4wdSBQboJcMPpHq7QKrN7rvvqp9e//jiv5WoIaK8uGqvriHosz/a07MPz77vSE6OXgrD0fr19vxZz+L10ftn76pf3/z8/qh6/+poXf3wH7D99SjrfQKyyz8PjhWO3l7fbM9ubw7ePn9efd7sKvw7Abdj9Wl3ebvFa9B+qbbtzXa4/NBqZxeX7fn1dnf423H99vdO4S+b3clUhU9bkfFEfby6uby9vNoVXtkKaCl1FbTOlugUeffL+qh68ern1y9Pjt5Ur37831H10xE+vf7l3ftVvfrjenv64aNsvlkd/vb7l38A)for you ;-)

Sometimes, when I try to unite a CompoundPath with a Path, some subpaths are "lost". 
Like here when I try to unite the CompoundPath in black with a Path in red. The resulting Path in blue is missing some subpaths.

![capture d ecran 2017-03-15 a 17 33 14](https://cloud.githubusercontent.com/assets/758717/23959495/ab3591e6-09a5-11e7-8a78-637126fe0205.png)

**Comment by lehni** [comment]
[Simplifed demo](http://sketch.paperjs.org/#S/nVldc922Ef0rmvtid0pigF0sgE0nT3mtZuz0Mc6DIt+OVTOSR5bViTr97z1LgDRJ0wqrJ14Ri4PFfp6l/nO6vfr9fPrh9I+P54frD6fudH333v5+vLq/eHPx48Xt+d8XP939/unuy+37N1cPH16/O12K74SiUyqRH3uO6gqTDKTimDhz14cszqsot5cpl66+izE81h3R09BvlnOQ/HSZmF2KqWMhR+QD4RCXKAf60Gcn3mt47EMSRxITMIp3hQILzk3JxYQ/2lvWxN388nHaYzAlaEiPE+70d0jRQcWYh81+JWnvFifZS9Ok7Xm6jCTQnEQ7UXKhEKeBXWBcDFtciSJxsGeR0mHBJ5ahNwnOqcNCDpGeLhkG8BRhycQO1oliKF5ZK0r2yVCSqoEwEVcQO6adApBgGkavXaLklIXLfLRJkwYdFrtSgUC/VA7nPV2GwC4XCbkTn1zGj3oDnDHCmNTQtOja7hWMnfN0SUTQgRNgcnBRNJU1jFliCWOWWsLYAmBKchSFSqfB+QCYeVM/3WGCne44XXbSzu4UXca9EXVwUyp5vIGp0E1HDs0ZawMvlIOzg3daQpFO1cH9NG4yFbrJG7Pbdm8EgadLIQ/1soUZI+BgzOnkpXVH3VbWbTesINALm5CQHaCc5qS0vlGLuxYxs1k2EeODaYA9JcDTWj0ymm4Zdsv4GPrFDe3qMO5o0lgQz3FMvBLWMN/LgQmm5oAgVxUJWBB3Bb9IaJ0EzUnL+FjB1DuRz46jpE6SFZMceZUEc/QuDLGyb00C9nB18MjrFLLLnkOFiT50U/oaDMTDbhJUN3FG4fMlw8DeJRFe58CfRwyOebrU4oJC3agohig+sr7RlEjP34hY4ZvIkAnFZVxklUizk6ZNueR1IiEY4Gv2joMUoBRGVgP8JTConuabRJ0UlAncZQMzRd7iCmuY6msuTmKAjMC+OaDerEwzhsgymQz222QSwaVyCqETmIbs57qIT9rg+Zw2Yg2NAnLSw9ZF/CaAN7aZk3KZ64gZ65DKqFUFwaM+5d0AXha4tbuBZh2FoIwGQkdJLuQXwsSCVJRMEbkNtdCoaRXBU0+ZI3i28CoP0GodvI3WJGjxqhTiOp1aIszptF86OWaX1PuISxWHza1Pbuv4shOtS0TtBriKosuiBOeI1EovAREfEXsI4g7x50QLgmYvLZcxstNtWY1LFHTbhO6Czk/rBjdrs82EtTYoeKjkRRjdls1GcXOpTQ3+9lJzh/OSzMBoDZG3RGRuty0TRr/1G/9bQgWQvJAJVYLwS8pu7H1TsVq3qpaJEYQqwxAoWHB7gJGO1b0VCm5q2YQboXI670FD1yViKsEtB7F5p3jCK06K2oXQm7gYN5zMsFdovlIri/AZRlhAQ0AoEXggfOolrNJgm01fYRaVB3fyZJxIO0KXQpMqqfpAOFUfNO0KlBmVQu5P630TQOSB0DIosNFFtEz0mubKkipM1Oo2PLv2fpjW+7aAO5WIuhCNzap1So2hqpwQwWMHaNQEz669H6b1vi1UmJTV89jl1BMN/fd2bVGn4+qdfCognQmFXH2hZr7qhByrr/Gshqrmz6EpEq1U5QIq7sFCIvKRkuj/DcFgeBIzBppsKT2G1P6eLWI7yaoLqHewaJKIEoV4a5QJvuyab3d9bOt9W6gtycghYIoVCEntNnvbtrDTeUbCs6WiNRNv4xsGqJfAIO4c2iPSEwkDhocLvkibgNLkbU5DOrGNKvtxK7nljOaduDVlxOJ/JIrJU0ovgQGRR2fLKBZqHSmWBpJr2OZS9nxty31bt4kARRI+hiocXUGJ0u+glJaJmXZgjH0XkCqUOgAin3UqHjvbtrDTeYDBdJuCoj7agBICFX6JNshitAArojAvhgJP4+wdff7QnnXIx4+FcESRLgnz9BFhAkHShNnwiDCjhBPySOdPACthvFiqAT+ApfhyRBh/2veQHA8hQ5ggfEgNRrNBQc35kDACECOp8CE1YDo4ho6pAXItCk5xSBj1igVT5SHTobNHEB86pLNYzyz5GDKEQVvlkOkIhUTg7ueQ0f5tDuSMsEPZ8ZTXAZ2nr1Zj2C2kk9XOkA8JGxsEdTsoHWvP1GOKsGTMohLTMeyIcsYg6sekg3E1SWllwFl6a0BIZ0iHY9JRneflt7vnpFH1YBNSOiQdbZxFVynHsCnYuKkH9QY2GEiQY5qAuUj2xMc0QfNSY13HpD1GdVDH56QjJjQ0kGATGjsG2aY24qKfjG0wVtaCZ9feD9N63xbQPFAqMBMpmjNgCgakaVJObVppA2RuM7Ax3Gm9cQHTRuBFb42YkMaCEbaxaR3ZH4O1Du3ZenX7PKOVU9qCfVEqtnkcQxDNkrOu2DSmgji05zy8rmg9FqzFw+3YDBKloJfQmBvDDfVzrNfpS2ZlDkkrjK33TcDIP3gtEVl7Fkuh9okx+MoK1WsjO9QM3O7oGy+EALhlCehMiETMi2gOyaeXoATrySFGYxwY1cCYZZhWuyY+7LnN1qfzoIx4KynBLIOhk6t5xRepdpk+BI4zkWD4Gqblvq0ba/EgUCOtR8kRjGpf00vsGwZq4jKqMfKwoIHui2tUCbP4u9Nf/vbu1v6z8fZ7/9kQ+1I+fvb9Bi4p2KoE/Xp6xft0f/ev8/WDu7p+uHk8//3qj/O9+3x9NZxfo012F7/47sL/aoLvbt+4f94Mw093w909FHj123B1/fEVVt6u39+f378a5U3Vn/Hmjftye/Nwfv3WYH7egnw5vxpff7r7fPNwc3fr/rj4648X0Xu8PXWn3+7PVx8/3d3cPnw+/fDLr//9Hw==)

**Comment by lehni** [comment]
@iconexperience I believe this is an issue for you. I used the `boolean-debug` branch to display the determined winding contributions, and see what it shows:

<img width="602" alt="screen shot 2017-03-18 at 13 37 56" src="https://cloud.githubusercontent.com/assets/213275/24072048/3830336c-0be0-11e7-84dd-7144a2cb2766.png">

Not surprisingly, the squares with winding 0 in all segments all disappear... But why do they get these values, while the others don't?

**Comment by lehni** [comment]
And with that information, here now an even more simplified case that has only one such square with wrong winding:

[Further simplified case](http://sketch.paperjs.org/#S/XZA7b4NAEIT/yukabAVW98bnyJUbN5bAKUMKDBeBwHcIA1ES5b+HVwqn2dWORt+M9hvb9GbwHr9UpssK7OPM5dM9pC2K0AFZ84GO7ta43uZR2hWbBJ+VIKAk9bkOQYVFQAVIIodASBCyXi6foLPcaQgJ9zmjIAUbOCgWFvMcgnmdVstXgrfPiY3gvazro6tdO0Z717o33ihPXeK1y18HKRVoTgXzudiB1JzQhU9ZoTQwKqleQyhb8fEjvjX5RF/4l1GJoLdlZzbxZL7875JmlTfrjbuXXeksfKKnA6KEjOr4uGtr0qpxpe3ueP/69vML)

**Comment by lehni** [comment]
@iconexperience so after looking a bit more into `getWinding()`, it turns out that this happens because of this line here:

https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L572
```js
// TODO: Determine how to handle quality when curve is crossed
// at starting point. Do we always need to set to 0?
quality = 0;
```

It is time to address that TODO now. I don't understand things enough in that part of the code to do it. Could you help out?

I've tried just removing it, but that produces one failing test. I've also tried moving this code outside of its condition block:

https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L545

```js
// Determine the quality of the winding calculation. Reduce the
// quality with every crossing of the ray very close to the
// path. This means that if the point is on or near multiple
// curves, the quality becomes less than 0.5.
if (a > pa - qualityEpsilon && a < pa + qualityEpsilon)
    quality /= 2;
```

That actually worked and produced no failing tests, but then the quality of all windings in that square is 0.25, which I don't think is right...

**Comment by lehni** [comment]
We should probably also look into why this happens on some of the squares and not on others? That's very bizarre...

**Comment by lehni** [comment]
@iconexperience I've spoken too quickly. By slightly rotating the square, I can the approach described above to produce wrong windings too:
[Updated Sketch](http://sketch.paperjs.org/#S/ZZHNboMwEIRfBflCqpIVxjaERDnl0kMjQXJsenDALS7ERsYhbaq+e/lrVTUXWzs73m9W/kSKnwRaon0pbFYgD2U67+uWGydx1o4SF2ejT7U+qzzhtpgd0DakPoQMeySOIIyKOabAfNbOKQPKqrHy/KuzZYsYIp94JMDAaNASCIOoGM52PlwPk+V6QHerg0qgsUaXYqMrbTq6e6zOwu0afZx0ivMTg7EQYoJp4BG6ABYTH48EHBRhDAFmOJ4wOJgA6X+AEbn7B7zPeCXVa9dJb5QXXjVi8Fa8sZtCVjkYbbkVM9zPHlPuOmcCZyU7Oe3l3e1OPCvdoVPrRlqpFbw792sHM38cI8UFMqGs6P210W8is8AzK1vxyD+E+X23msxXrU+dlay6DzwawctaS2UbtHx6/voG)

The problem here is that I get one winding value of 0 with quality of 0.5...

**Comment by iconexperience** [comment]
@lehni The quality factor should not be relevant in this case, because it is meant to handle cases where two curves are very close and detection of intersections is not reliable. It's quite possible that the quality factor causes this bug, but that only shows that the implementation is not correct.

In my opinion in the example we should get a correct winding result for all tested points. If that was the case, the quality factor should be irrelevant. But I agree, we should urgently settle this, or at least have a clear documentation for the quality factor.

But first I will try to find out why some points give an incorrect winding result.

**Comment by iconexperience** [comment]
Alright, this is the old problem with corners. Here is an example where the winding calculates incorrectly:

![image](https://cloud.githubusercontent.com/assets/2122773/24084275/421c20ea-0ce7-11e7-9ce8-b730fb34ed80.png)

The grey line is the vertical ray. The winding left and right (here up and down) are zero, which is correct. But we need a special handling of these cases. For `contains()` we check if the point is on the path in these cases, but this will not be sufficient if we need the correct winding number.

I think long time ago we simple assinged `1` to these cases, but this is not always correct, because if you have two identical squares, the winding at that point must be `2`. And if there are two identical squares but with opposite orientation, the winding must be `0`. It's not that simple.

So how to determine the correct winding in these cases? I think by calculating the "on-path winding". This can still lead to incorrect results if the overlapping curves are not fully identical, but those cases should get a bad quality factor.

Welcome back "on-path winding"!

**Comment by iconexperience** [comment]
BTW, this is a perfect example to show why I selected odd values for `propagateWindings()`. If you check the windings at time values  `0.25`, `0.5` and `0.75` of a square, you will always check at a corner.  This will not happen as easily if you use random looking values like `0.2341`, `0.48732` and `0.76132`.

**Comment by lehni** [comment]
@iconexperience thanks for looking into this! I agree with you, but I think we need an implementation that can handle these offsets because there will always be a shape that will be able to trigger this error otherwise. I was basically hoping to hit these errors eventually that you were addressing by shifting the factors : )

**Comment by iconexperience** [comment]
Yes, I agree, this should be our goal. There are several approaches that I thought of, but all have their problems.

1. If we encounter a corner, try again by casting the ray along the other axis (vertically instead of horizontally or vice versa). This approach would solve this issue, but fails if the squares are not rotated. Also, it would fails for corners with small angles, like the following case, which fails for both axes:

![image](https://cloud.githubusercontent.com/assets/2122773/24085269/0cd54b20-0cf9-11e7-8731-c34d97b3da8a.png)

2. Add the winding to left winding and right winding if we are on a path. This does not help in this example, because we will be on the path twice and the windings would cancel each other. 

3. Use `0.25`, `0.5` and `0.75`, but if we happen to be at a segment (`t < winding epsilon || t > 1 - windingEpsilon`) we will move the time value away from the segment. This may be the best approach. There could still be problems with corner with very small angles, but most of those would get a bad quality factor. Still no guarantee for a correct result, but the number of wrong winding values should be much, much smaller.

**Comment by lehni** [comment]
@iconexperience regarding 3.:

Aren't we already doing this in https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L718 ?

Your'e proposing to use `windingEpsilon`, but that's not for the curve-time space... And it's actually smaller than `CURVETIME_EPSILON`.

**Comment by lehni** [comment]
@iconexperience I've just brought the on-path winding handling back, and it looks like together with another change I made earlier, both rotated and not rotated squares are now handled correctly. I haven't addressed 1. & 3. of your suggestions above yet at all.

<img width="644" alt="screen shot 2017-03-20 at 01 05 24" src="https://cloud.githubusercontent.com/assets/213275/24086115/79c3cc74-0d09-11e7-9fb6-9927f6879d8e.png">

**Comment by lehni** [comment]
@iconexperience: Regarding suggestion 3: The problem is that we'd need to physically be `windingEpsilon` away from the segment (either horizontally or vertically, depending on 
`dir`), but that does not translate to the same amount in curve-time, and there is no quick way translate one to the other. Do you think this is still required now that the on-path winding is back? I am not sure if I can close this issue after implementing a unit test for it, or if there is more to do now?

Regarding 1: How would we know we're on a corner inside `getWinding()`?

**Comment by lehni** [comment]
This has been long fixed, since around `v0.11.0`
